bin/generate_peptides.py

```diff
@@ -69,59 +69,125 @@ def parse_args(args=None):
+
     print("# proteins: ", len(protid_protseq_protlen))
 
     # write out protein lengths
     protid_protseq_protlen[["protein_id", "protein_length"]].to_csv(args.proteins_lengths, sep="\t", index=False)
 
     ####################
     # generate peptides
-
     with gzip.open(args.peptides, "wt") as pep_handle:
-        print("peptide_id", "peptide_sequence", sep="\t", file=pep_handle, flush=True)
-        print("protein_id", "peptide_id", "count", sep="\t", file=args.proteins_peptides, flush=True)
+        print_header = True
         id_counter = 0
+
         # for each k
         for k in range(args.min_len, args.max_len + 1):
             print("Generate peptides of length ", k, " ...", flush=True)
-            # for each protein generate all peptides of length k
-
-            results = pd.DataFrame(
-                [
-                    (str(it.protein_id), pep)
-                    for it in protid_protseq_protlen.itertuples()
-                    for pep in gen_peptides(it.protein_sequence, k)
-                ],
-                columns=["protein_id", "peptide_sequence"],
-            )
-
-            # TODO handle this more memory efficiently
-            print("format results ...", flush=True)
-            # count occurences of one peptide in one protein
-            results = results.groupby(["protein_id", "peptide_sequence"]).size().reset_index(name="count")
-
-            unique_peptides = results[["peptide_sequence"]].drop_duplicates()
-            unique_peptides["peptide_id"] = range(id_counter, id_counter + len(unique_peptides))
-            id_counter += len(unique_peptides)
-            unique_peptides[["peptide_id", "peptide_sequence"]].to_csv(
-                pep_handle, mode="a", sep="\t", index=False, header=False
-            )
-
-            results = results.merge(unique_peptides)
-            results[["protein_id", "peptide_id", "count"]].to_csv(
-                args.proteins_peptides, mode="a", sep="\t", index=False, header=False
-            )
-
-            print("# peptides of length ", k, ", (non-unique across proteins): ", len(results))
+
+            # Note: could be done with prefixes instead of single first letters if this remains bottleneck
+            for prefix in aa_list:
+                print("with prefix ", prefix, flush=True)
+                # for each protein generate all peptides of length k with current prefix (to reduce peak mem usage)
+                # (the AA-wise processing causes multiple iterations over the same protein sequences,
+                # but the increase of run time is negligible in this context)
+                results = pd.DataFrame(
+                    [
+                        (str(it.protein_id), pep)
+                        for it in protid_protseq_protlen.itertuples()
+                        for pep in gen_peptides(it.protein_sequence, k, prefix)
+                    ],
+                    columns=["protein_id", "peptide_sequence"],
+                )
```

Comment:
```
results = pd.DataFrame(
                     [
                         (str(it.protein_id), pep)
                         for it in protid_protseq_protlen.itertuples()
                         for pep in gen_peptides(it.protein_sequence, k, prefix)
                     ],
                     columns=["protein_id", "peptide_sequence"],
                 )
```

Instead of adding everything in this list you could use a set wrapping to only add unique tuples of peptide_sequence and protein_id. To not lose the information of how many peptides were added you can add a third column directly here (As you preselected the peptides by the first amino acid there will be no other count of this peptide in the same protein). I think you could save some resources by skipping the grouping of the results df and making it unique later on.

Apart from that there is one "," after columns which can be deleted.

Reply:
```
protein_tuples = [
                    (str(it.protein_id), pep)
                    for it in protid_protseq_protlen.itertuples()
                    for pep in gen_peptides(it.protein_sequence, k, prefix)
                    ]
results = pd.DataFrame(set([
                        (prot_id, pep_seq, protein_tuples.count((prot_id, pep_seq)))
                            for prot_id, pep_seq in protein_tuples
                        ]),
                     columns=["protein_id", "peptide_sequence", "count"]
                 )
```

I had something like this in mind

Reply:
The "," was for some reason introduced by the python Black formatter

Reply:
"You might have noticed that closing brackets are always dedented and that a trailing comma is always added. Such formatting produces smaller diffs; when you add or remove an element, it’s always just one line. " taken from: https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html

So I guess that's more efficient for changes!

Reply:
seems the current version with `groupBy().size()` can not really be optimised ...